_posts/2016-03-10-information-hiding.md

```diff
@@ -0,0 +1,334 @@
+---
+layout: post
+title: Information hiding, enforced
+
+meta:
+  nav: blog
+  author: adelbertc
+  pygments: true
+---
+
+Code should be reusable. An expression traversing a data structure
+shouldn't be written multiple times, it should be pulled out into a
+generic traversal function. At a larger scale, a random number generator
+shouldn't be written multiple times, but rather pulled out into a
+module that can be used by others.
+
+It is important that such abstractions must be done carefully.
+Often times a type is visible to the caller, and if the type
+is not handled carefully the abstraction can leak.
+
+For example, a set with fast random indexing (useful for random
+walks on a graph) can be implemented with a sorted `Vector`.
+However, if the `Vector` type is
+leaked, the user can use this knowledge to violate the invariant.
+
+```scala
+import scala.annotation.tailrec
+
+def binarySearch(i: Int, repr: Vector[Int]): (Boolean, Int) = {
```

Comment:
So one reason I left it on is because I'm running this all through `tut` so the code needs to be present to "pass." I do agree it's clutter though so I'll remove it after running it through tut